pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -152,14 +246,18 @@ func buildClusterKey(service *model.Service, port *model.Port, cb *ClusterBuilde
 
 // buildOutboundClusters generates all outbound (including subsets) clusters for a given proxy.
 func (configgen *ConfigGeneratorImpl) buildOutboundClusters(cb *ClusterBuilder, cp clusterPatcher) ([]*discovery.Resource, cacheStats) {
-	resources := make([]*discovery.Resource, 0)
-	hit, miss := 0, 0
 	var services []*model.Service
 	if features.FilterGatewayClusterConfig && cb.proxy.Type == model.Router {
 		services = cb.push.GatewayServices(cb.proxy)
 	} else {
 		services = cb.push.Services(cb.proxy)
 	}
+	return configgen.buildOutboundClustersWithServices(cb, cp, services)
+}
+
+func (configgen *ConfigGeneratorImpl) buildOutboundClustersWithServices(cb *ClusterBuilder, cp clusterPatcher, services []*model.Service) ([]*discovery.Resource, cacheStats) {
```

Comment:
Any specific reason why this has been refactored to separate method? I am  fine with this but just curious about the reason? Do we intend to reuse some where?

Reply:
yes. the default implementation (`buildOutboundClusters`), would build all outbound clusters, using all the services (see https://github.com/istio/istio/blob/master/pilot/pkg/networking/core/v1alpha3/cluster.go#L161). For delta, we need to generate only based on a set of services, which are defined as which services changed. So, I needed a way to pass in the services based on which it generates the clusters.